commonmark/src/main/java/org/commonmark/internal/inline/Scanner.java

```diff
@@ -0,0 +1,187 @@
+package org.commonmark.internal.inline;
+
+import org.commonmark.internal.util.CharMatcher;
+
+import java.util.List;
+
+public class Scanner {
+
+    // Lines without newlines at the end. The scanner will yield `\n` between lines because they're significant for
+    // parsing and the final output. There is no `\n` after the last line.
+    private final List<CharSequence> lines;
+    // Which line we're at.
+    private int lineIndex;
+    // The index within the line. If index == length(), we pretend that there's a `\n` and only advance after we yield
+    // that.
+    private int index;
+
+    // Current line or "" if at the end of the lines (using "" instead of null saves a null check)
+    private CharSequence line = "";
+    private int lineLength = 0;
+
+    // TODO: Visibility
+    public Scanner(List<CharSequence> lines, int lineIndex, int index) {
+        this.lines = lines;
+        this.lineIndex = lineIndex;
+        this.index = index;
+        if (!lines.isEmpty()) {
+            line = lines.get(lineIndex);
+            lineLength = line.length();
+        }
+    }
+
+    public char peek() {
+        if (index < lineLength) {
+            return line.charAt(index);
+        } else {
+            if (lineIndex < lines.size() - 1) {
+                return '\n';
+            } else {
+                // Don't return newline for end of last line
+                return '\0';
+            }
+        }
+    }
+
+    public char peekPrevious() {
+        if (index > 0) {
+            int prev = index - 1;
+            return line.charAt(prev);
+        } else {
+            if (lineIndex > 0) {
+                return '\n';
+            } else {
+                return '\0';
+            }
+        }
+    }
+
+    public boolean hasNext() {
+        if (index < lineLength) {
+            return true;
+        } else {
+            // No newline at end of last line
+            return lineIndex < lines.size() - 1;
+        }
+    }
+
+    public void next() {
+        index++;
+        if (index > lineLength) {
+            lineIndex++;
+            if (lineIndex < lines.size()) {
+                line = lines.get(lineIndex);
+                lineLength = line.length();
+            } else {
+                line = "";
+                lineLength = 0;
+            }
+            index = 0;
+        }
+    }
+
+    public boolean next(char c) {
```

Comment:
While reading the code this `next` method makes me think a different action for this method. When I renamed to `isNext(...)` it more senses to me. what do you think?

Reply:
Yeah, it's really a `nextIf`. What I like about `next` is that it reads well in code like `if (scanner.next('x'))`.

With `isNext` it's not clear that it advances the scanner state to me.